enterprise/models.py

```diff
@@ -321,6 +368,147 @@ def __repr__(self):
         return self.__str__()
 
 
+class UserDataSharingConsentAudit(TimeStampedModel):
+    """
+    Object that exists to store the canonical state of whether a particular
+    user has given consent for their course data to be shared with a particular
+    single-sign-on identity provider
+    """
+    class Meta(object):
+        app_label = 'enterprise'
+        verbose_name = "Data Sharing Consent Audit State"
+        verbose_name_plural = "Data Sharing Consent Audit States"
+        unique_together = ('user', 'enterprise_customer',)
+
+    NOT_SET = 'not_set'
+    ENABLED = 'enabled'
+    DISABLED = 'disabled'
+    STATE_CHOICES = (
+        (NOT_SET, 'Not set'),
+        (ENABLED, 'Enabled'),
+        (DISABLED, 'Disabled'),
+    )
+
+    user = models.ForeignKey(
+        User,
+        on_delete=models.CASCADE,
+    )
+
+    enterprise_customer = models.ForeignKey(
+        EnterpriseCustomer,
+        on_delete=models.CASCADE,
+    )
+
+    state = models.CharField(
+        max_length=8,
+        blank=False,
+        choices=STATE_CHOICES,
+        default=NOT_SET,
+        help_text=_(
+            "Stores whether the user linked to this model has consented to have "
+            "their information shared with the linked EnterpriseCustomer."
+        )
+    )
+
+    @property
+    def enabled(self):
+        """
+        Determine whether the user has enabled data sharing
+        """
+        return self.state == self.ENABLED
+
+    def enable(self):
+        """
+        If data sharing is not already enabled, enable it and create a new
+        historical record indicating the change.
+        """
+        if self.enabled:
+            return
+        self.historical_changes.create(
+            previous_state=self.state,
+            new_state=self.ENABLED,
+        )
+        self.state = self.ENABLED
+
+    def disable(self):
+        """
+        If data sharing is not disabled, disable it and create a new
+        historical record indicating the change.
+        """
+        if self.state == self.DISABLED:
+            # Note we don't call self.enabled; we want to set to disabled state
+            # explicitly if state is not set yet, which shows up as disabled
+            return
+        self.historical_changes.create(
+            previous_state=self.state,
+            new_state=self.DISABLED,
+        )
+        self.state = self.DISABLED
+
+    def __unicode__(self):
+        return 'UserDataSharingConsent: {} - {} - {}'.format(
+            self.user,
+            self.enterprise_customer.name,
+            self.state,
+        )
+
+
+class UserDataSharingConsentAuditHistory(TimeStampedModel):
```

Comment:
We weren't using it on the feature branch for TPA, of which this is basically a clone, but I can look into whether that's an option here.